Design proposal for training

mvsusp opened the pull request:
# Amazon SageMaker Containers

**SageMaker Containers** is the open source library that the SageMaker team uses to create the open source containers
that run on [Amazon SageMaker](https://aws.amazon.com/documentation/sagemaker/):

- https://github.com/aws/sagemaker-tensorflow-containers
- https://github.com/aws/sagemaker-mxnet-containers
    
 You can use SageMaker Containers as a **library** with helper functions for your training script or to 
 **create** your own ML framework container compatible with SageMaker. 
 
## Install
 
### Installing from source
 ```bash
pip install -U .
```

## Getting started
### Using it as a **library**
```sagemaker_containers.training_environment``` contains information about the training container including: 
hyperparameters, instance information (number of CPUs, GPUs, host name), input data information, environment
variables, etc.

Let's suppose that you want to train the Keras script below in SageMaker:
```my_training_script.py```
```python
import argparse
import os

import numpy as np
import tensorflow as tf

parser = argparse.ArgumentParser()
parser.add_argument('--training-data-dir', type=str)
parser.add_argument('--batch-size', type=int)
parser.add_argument('--model-dir', type=str)

args = parser.parse_args()

data = np.load(os.path.join(args.training_data_dir, 'training_data.npz'))
x_train, y_train = data['features'], keras.utils.to_categorical(data['labels'])

model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(64, activation='relu', input_dim=20))
model.add(tf.keras.layers.Dropout(0.5))

sgd = tf.keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])

model.fit(x_train, y_train, epochs=20, batch_size=args.batch_size)

# saves the model in the end of training
model.save(os.path.join(args.model_dir, 'saved_model.h5'))
```

To be able to train the script above in SageMaker, you need to provide the parameters ```--training-data-dir```,
```--batch-size```, ```--model-dir```. ```training_environment``` makes this process very easy:

```python
import sagemaker_containers.training_environment as env
...

# TODO (mvsusp) - create env.training_data_dir and env.evaluation_data_dir helpers
parser.add_argument('--training-data-dir', type=str, default=env.training_data_dir)
parser.add_argument('--batch-size', type=int, default=env.batch_size)
parser.add_argument('--model-dir', type=str, default=env.model_dir)
```

We can use [SageMaker Python SDK](https://github.com/aws/sagemaker-python-sdk) to submit ```my_training_script.py``` to
trainining using the ```TensorFlow``` estimator:

```python
# TODO (mvsusp) - allow sagemaker-python-sdk estimator to work in the scenario below
from sagemaker.tensorflow import TensorFlow

estimator = TensorFlow('train.py', role='SageMakerRole', 
                       train_instance_type='ml.p2.xlarge', train_instance_count = 1)
estimator.fit({'training': 's3://my/bucket/with/the/training/data'})
```

### Creating a new ML framework container compatible with SageMaker Training
The core training structure of SageMaker Containers is ```Training```. When training starts, e.g. 
`training.run()`, the `Training` instance executes the following steps:

1. install any required python dependencies
2. start customs metrics that will be reported to CloudWatch Logs
3. load SageMaker Training environment information, including default configuration files, directories, 
    environment variables, and hyperparameters
4. download the user script (or Python package) containing the functions required by the framework
5. start the training process
6. report success/failure
  
Let's suppose that we want to create a Keras Container that trains any Keras model provided by the **user script**.

Let's suppose that to be able to use our Containers, users need to submit a Python script containing a function with 
the following signature:  
```python
def train(training_dir, hyperparameters):
    """Return a Keras model that will be trained by our SageMaker Keras container.
    
    Args:
        - training_dir (string): the path to the directory containing the training 
            data provided by the user.
        - hyperparameters (dict[string, object]): a map containing the hyperparameters 
            provided by the user.
         
    Returns:
        - (Keras.Model): the model that will be saved in the end of the training.
    """
```

#### Implementing our Keras Framework Container
```Training``` will install the user provided script as a module, and prepare the training environment for us. All we 
have to do is implement a function, ```training_process_fn```, which implements the training process of our framework.

```training_process_fn``` takes as parameters **user module** and the **training environment**, and is responsible to 
save the trained model in end of training:
```python
import os

from sagemaker_containers import Training

def keras_framework_training_process(user_module, training_environment):
    """This function implements our Keras Framework.
    
    Args:
        - user_module (module): the python script provided by the user loaded as a Python ```module```
        - training_environment (sagemaker_containers.TrainingEnvironment): the container's 
            training environment
    """
    # retrieves the training dir from training channels
    training_dir = training_environment.channel_input_dirs['training']
    
    # retrieves the hyperparameters submitted by the user
    hps = training_environment.hyperparameters
    
    # invokes user's ```train``` function
    model = user_module.train(training_dir, hps)
    
    # saves the model in the end of training
    model.save(os.path.join(training_environment.model_dir, 'saved_model.h5'))

# Creates the training from the training process function and start training.
Training.from_train_fn(train_fn=keras_framework_training_process).run()
```

#TODO (mvsusp) - complete the example after implementing support for Docker containers.

#### An example of user provided script for training:
__train.py__
```python
import os

import keras
import numpy as np

def train(training_dir, hyperparameters):
    data = np.load(os.path.join(training_dir, hyperparameters['training_data_file']))
    x_train, y_train = data['features'], keras.utils.to_categorical(data['labels'])

    model = keras.models.Sequential()
    model.add(keras.layers.Dense(64, activation='relu', input_dim=20))
    model.add(keras.layers.Dropout(0.5))

    sgd = keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
    model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])

    model.fit(x_train, y_train, epochs=20, batch_size=128)
    return model
```
#### Submitting a Python script for training using our Keras container
We can use [SageMaker Python SDK](https://github.com/aws/sagemaker-python-sdk) to submit a scripts to our 
containers:

```python
# TODO (mvsusp) - allow sagemaker-python-sdk estimator to work in the scenario below
from sagemaker.estimator import Estimator

estimator = Estimator('train.py', role='SageMakerRole', 
                      train_instance_type='ml.p2.xlarge', train_instance_count = 1)
estimator.fit({'training': 's3://my/bucket/with/the/training/data'})
```

You can look at more information on how to use ```SageMaker Python SDK``` here: https://github.com/aws/sagemaker-python-sdk#byo-docker-containers-with-sagemaker-estimators.

mvsusp committed:
Add App, TrainingEngine, TrainingEnvironment

mvsusp committed:
Remove non-necessary code

mvsusp committed:
Design update

jesterhazy commented:
- framework code expects user code to return a model, but example user code doesn't return. actually user code isn't a function at all. 
- mxnet containers use return value from train to decide whether to call save function, and that function expects the value to be a Module/gluon network
- save_model treats model as a dict... but this would usually be a model/network of some kind, at least for mxnet users
- why do we need to set 'saved' key on model at all?

mvsusp commented:
About @jesterhazy considerations:

1. user code actually returns a model. The user code example mentioned is:
```python 
def train(chanel_dirs, hps): 
    return {'trained': True, 'saved': False}
```
2. you are correct that ```train``` should return model. That is not going to change. I will add more documentation and design pictures to explain better
3. saving a model is framework specific. The example above is just a mock ML framework that saves a dict as a model
4. that is framework specific as well. I will add another example in the docs to clarify your doubts.

jesterhazy commented:
i was looking at `training_module/train.py` under User Module. what's the relation between that and user script?

mvsusp commented:
@jesterhazy  
Users can provide a ```def train(...):``` that returns a model.
In addition to that, users can provide a python package, that will be executed in the container with the follow command: ```python -m {module_name}```. 
The latter option allows users to execute any arbitrary script they want in the container.
Customers can use the library API in their own scripts (freedom to opt-in and use the methods that they want).

iquintero commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class Training(object):
+    def __init__(self, train_fn=None, module_name=None):
```
` train_fn` and `module_name` are mutually exclusive. (As inferred from the existance of  `from_train_fn()` and `from_module()`. This class needs to throw an exception if it is manually constructed and passed both.

mvsusp replied:
@iquintero, I refactored the class to cover that scenario: https://github.com/mvsusp/sagemaker-container-support/pull/3/files#diff-35452821e8fa0a4ef79ea98492563a4aR35

Please, let me know your thoughts.

iquintero commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,54 @@
+            raise e
+
+    def import_user_module(self, code_dir, user_script='user_script'):
+        return importlib.import_module(user_script)
```
`code_dir` is not used, but it probably should right?

mvsusp replied:
refactored out this function: https://github.com/mvsusp/sagemaker-container-support/pull/3/files#diff-35452821e8fa0a4ef79ea98492563a4aR81

Let me know your thoughts.

iquintero commented on src/sagemaker_containers/training_environment.py:
```diff
@@ -0,0 +1,82 @@
+    @property
+    def input_config_dir(self):
+        """The directory where standard SageMaker configuration files are located."""
+        return os.path.join(self.base_dir, 'input/config')
```
I would rewrite this as

```python
os.path.join(self.base_dir, 'input', 'config')
```

while I think os.path.join() will figure out the / in your string correctly (even on Windows), Windows only users can be confused as to why input/config is a path.

mvsusp replied:
done

iquintero commented on src/sagemaker_containers/environment.py:
```diff
@@ -0,0 +1,49 @@
+
+    @property
+    def model_dir(self):
+        return os.path.join(self.base_dir, 'model')
```
Not a big fan of having the getter function of the property calculate its value every time. Can we move to having:

```python
def __init__(self):
    self._model_dir = os.path.join(self.base_dir, 'model')
     # ... other stuff here

@property
def model_dir(self):
    return self._model_dir
```

If someone else is strongly opposed to this Im fine discussing it though

mvsusp replied:
moved all logic to the constructor.

iquintero commented on test/functional/test_training.py:
```diff
@@ -0,0 +1,91 @@
+    return tmpdir
+
+
+def test_training_from_train_fn(opt_ml, model_file):
```
Add a test to ensure that your class handles both a train_fn and a train_module being passed at the same time.

mvsusp replied:
refactored the classes so we do not have the edge case anymore.

iquintero commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,54 @@
+
+                self.train_fn(user_module, self.environment)
+            else:
+                subprocess.Popen(['python', '-m', self.module_name])
```
this asumes that the module is importing this as a library in order to figure out where to put stuff right? or the user just hardcoded all their paths but we don't care how they do it as long as it works right?

mvsusp replied:
When the user provides a module, we will validate it before starting the training job.
The module will be saved to the code_dir. When a new Environment instance is created, it adds the code dir to PYTHONPATH.

iquintero reviewed (commented):
the overall design makes sense to me, just have some comments.

mvsusp commented:
@iquintero thanks for the observations. The code here was simplified for design purposes. All your comments will be addressed in the following PRs with the real code implementation. Or would you prefer to address the changes here?

mvsusp commented:
@iquintero maybe I should not even push this PR since the code is not a real implementation?

laurenyu commented on src/sagemaker_containers/environment.py:
```diff
@@ -0,0 +1,49 @@
+    @property
+    def code_dir(self):
+        """The directory where user-supplied code will be staged."""
+        return os.path.join(self.base_dir, 'code')
```
'model' and 'code' can be constants with 'opt/ml'

mvsusp replied:
created constants with just the base_names

laurenyu commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,54 @@
+            self.environment.write_failure_file('Uncaught exception during training: %s' % e)
+            raise e
+
+    def import_user_module(self, code_dir, user_script='user_script'):
```
should we be giving a default value to `user_script`? It seems a little strange to me to presume that, especially if `code_dir` is required. I'd be more inclined to follow the strategy in PythonSDK, where `entry_point` is required and `source_dir` defaults to `None`

mvsusp replied:
Yeah, I agree that we should not. Just did that to reduce the prototype amount of lines and focus in the overall design.

laurenyu commented on src/sagemaker_containers/training_environment.py:
```diff
@@ -0,0 +1,82 @@
+
+    @property
+    def channel_dirs(self):
+        """"""
```
Why bother including the docstring at this point? 😂

laurenyu reviewed (commented):
are you planning a future PR that fleshes out the documentation in the README? (I'm guessing the current proposed README is just for showing what you're thinking)

also, my usual comments:
* capitalize "Python"
* import statements - I usually just ctrl/cmd-F on the [Python Standard Library list](https://docs.python.org/3/library/index.html) to check what goes in the first group.

mvsusp commented:
@laurenyu I agree with all your points here. I'm planning to not push this PR and just use it to gather design feedback.

owen-t commented on src/sagemaker_containers/environment.py:
```diff
@@ -0,0 +1,49 @@
+
+import sys
+
+base_dir = 'opt/ml'
```
Should this be /opt/ml ?

mvsusp replied:
fixed.

owen-t commented on src/sagemaker_containers/environment.py:
```diff
@@ -0,0 +1,49 @@
+
+import multiprocessing
+import os
+
```
whitespace?

mvsusp replied:
added flake8 so we won't have these error again.

owen-t commented on src/sagemaker_containers/environment.py:
```diff
@@ -0,0 +1,49 @@
+    @property
+    def available_gpus(self):
+        """The number of gpus available in the current container."""
+        pass
```
Why not provide a default here?

Try to invoke nvidia-smi (per Nacho's comment, probably not in the getter) and return the result or 0.

mvsusp replied:
done

owen-t commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class Training(object):
+    def __init__(self, train_fn=None, module_name=None):
```
API docs! Particularly so for this class as it's a little esoteric.

mvsusp replied:
@owen-t I added docstrings to the methods.

owen-t commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,54 @@
+
+class Training(object):
+    def __init__(self, train_fn=None, module_name=None):
+        self.environment = env.TrainingEnvironment()
```
Document the constraints on these arguments

mvsusp replied:
done

owen-t commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,54 @@
+
+                self.train_fn(user_module, self.environment)
+            else:
+                subprocess.Popen(['python', '-m', self.module_name])
```
Popen returns immediately and successfully if the process can be created. Should this be a check_call?

mvsusp replied:
That's correct @owen-t. Using check call. This functional tests shows the functionality:
https://github.com/mvsusp/sagemaker-container-support/pull/3/files#diff-82c8c20785ab069f8c8d755dc15cb636R49

owen-t commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,54 @@
+        try:
+            if self.train_fn:
+
+                user_module = self.import_user_module(self.environment.code_dir)
```
Should we run the train_fn in a subprocess as well?

owen-t commented on src/sagemaker_containers/training_environment.py:
```diff
@@ -0,0 +1,82 @@
+        return os.path.join(self.base_dir, 'output')
+
+    @property
+    def resource_config(self):
```
Can't we provide defaults for many of these?

owen-t commented on src/sagemaker_containers/training_environment.py:
```diff
@@ -0,0 +1,82 @@
+        """"""
+        pass
+
+    def load_training_parameters(self, fn):
```
Provide doc strings for these undocumented functions

mvsusp replied:
done

owen-t reviewed (commented):
Looks good - some of the implementations are missing though - let's get them filled out!

Great stuff!

mvsusp committed:
Add docstrings and handle PR comments

mvsusp committed:
Update Readme.md and small grammar mistakes

mvsusp committed:
Fix minor grammar mistakes

mvsusp committed:
Update README.md

mvsusp commented:
@jesterhazy I update the documentation and examples. Please let me know if you still have any
doubts/observations.

mvsusp committed:
Merge branch 'mvs/mvs-add-setup' into mvs-design-decisions-for-training

mvsusp commented:
@laurenyu I've found this flake8 extension that inspects the import order for us, so I stop messing up the import order: https://github.com/mvsusp/sagemaker-container-support/pull/3/files#diff-b91f3d5bd63fcd17221b267e851608e8R52


🎉 🍷

mvsusp committed:
Add constants

mvsusp commented:
@jesterhazy @owen-t @iquintero @laurenyu I updated the readme and added tox to this PR.
The next steps will be:

- proposal design for training 
- setting up team city
- additional missing features

jesterhazy commented on src/sagemaker_containers/environment.py:
```diff
@@ -0,0 +1,109 @@
+
+
+# TODO (mvsusp) -  create a shortcut decorator to allow get/set outside the class, e.g. env.model_dir = 'x'
+base_dir = BASE_PATH
```
why are we exposing as model var at all? it's already a class attribute.

mvsusp replied:
It is temporary, just to reduce the size of the PR. But I added a TODO in the top to fix it.

jesterhazy commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,138 @@
+        self.environment = env.TrainingEnvironment()
+        self.training_process_fn = training_process_fn
+
+    # TODO(mvsusp) - add update_training_environment_fn to allow frameworks to update the training env.
```
what does update mean? user can just update/attributes to the instance without adding a special function (b/c python)

mvsusp replied:
for the following scenario:

```python
    
def framework_train_fn(user_module, training_env):
    model = user_module.train(training_env.channel_input_dirs, training_env.hyperparameters)
    save_model(model, model_file)

def update_training_environment_fn(training_env):
    training_env.world_size = len(training_env.hosts * training_env.num_gpus)
    training_env.rank = 2
    return training_env


Training.from_train_fn(train_fn=framework_train_fn,  
                                      environment_fn=update_training_environment_fn)
```

Put it is outside the scope this PR. We can discuss implementation details later.

jesterhazy commented on test/functional/test_training.py:
```diff
@@ -0,0 +1,101 @@
+
+
+def write_user_module(tmpdir, code):
+    customer_module = tmpdir.mkdir('code', 'customer_module')
```
customer_module*_path* etc.

jesterhazy reviewed (commented):
looks good. added some nits. i think the readme docs will need another full pass once the design is settled.

winstonaws commented on README.md:
```diff
@@ -1,7 +1,188 @@
+...
+
+# TODO (mvsusp) - create env.training_data_dir and env.evaluation_data_dir helpers
+parser.add_argument('--training-data-dir', type=str, default=env.training_data_dir)
```
Does this actually work with the code as is right now?

the training_environment module doesn't have training_data_dir, batch_size, etc. fields. You have to create an instance of TrainingEnvironment, right?

It's worth thinking about whether there's value in needing to create an instance of the class. The functions could all just live in the module rather than the class. Is there a strong reason one way or the other currently?

winstonaws commented on src/sagemaker_containers/training.py:
```diff
@@ -0,0 +1,138 @@
+_PYTHONPATH = 'PYTHONPATH'
+
+
+class Training(object):
```
Two questions:

1). Can we design this so we don't lock users into this "framework"? I want to avoid the same problem we had with our original TensorFlow framework design. 

What would it look like if framework authors were implementing their train entrypoint using library functions that we provide? How much complexity would remain that we would still want to abstract away, at the cost of flexibility? I'd like to see us design that "API" first, and see some example user code in that style.

We may still want to offer something like this, which enforces a specific workflow for the entirety of training, which lets us ensure that all of our frameworks behave the same way. Even in that case, I'd argue that that should be written purely by calling those library functions.

2). For the "framework", is a class like this the best implementation? Other potential alternatives:
- A function which takes in a function/module
- A class which can be subclassed

Are there reasons for going this direction over the others?
